### ISSUE TITLE
Strange behavior of Sidekiq Pro's Reliability feature

### ISSUE THREAD

**Issue by mkaschenko** [mperham/sidekiq#2120]
Hello, @mperham. Thank you for the gem!

I face strange behavior of Sidekiq Pro's Reliability feature [1].

I have TestWorker [2]. Also I use 'weighted random' fetch algorithm [3].
I run sidekiq the same way as in [3] and push the worker (TestWorker.perform_async).
I kill the sidekiq process (kill -9 pid).
I run sidekiq again.

And now 2 TestWorkers run instead of 1.
If I change the weight `sidekiq ... -q default,2 ...` to `sidekiq ... -q default,5 ...`, then 5 TestWorkers run and so on.

What do you think about it?

[1] https://github.com/mperham/sidekiq/wiki/Reliability
[2] https://gist.github.com/mkaschenko/24da18c795962da3c970
[3] https://github.com/mperham/sidekiq/wiki/Reliability#weighted-random-algorithm

**Comment by mperham** [comment]
Wow, good catch!  You're right, this is a bug.

**Comment by jonhyman** [comment]
Wow. Thanks for reporting and thanks for fixing, @mperham!

**Comment by mperham** [comment]
Thank you for the great report, 1.9.2 has been released.

**Comment by mkaschenko** [comment]
:+1:

**Comment by parhamr** [comment]
@mperham I appreciate this change! Since upgrading from `1.9.1` to `1.9.2` I have seen a significant performance regression. Do you have tests that can check this or would you like some diagnostic data?

**Comment by mperham** [comment]
Can you give more versions: Ruby, sidekiq, sidekiq pro?  Ruby 1.9 support is pretty frozen at this point, any possibility of upgrading Ruby and sidekiq?

**Comment by parhamr** [comment]
Sorry, I was unclear: it was an upgrade from `sidekiq-pro 1.9.1` to `sidekiq-pro 1.9.2`.

Ruby version: `2.1.5 x86_64-linux 2014-11-13`
Rails version: `3.2.21`
Current Sidekiq versions in Gemfile.lock:
```
    sidekiq (3.2.6)
      celluloid (= 0.15.2)
      connection_pool (>= 2.0.0)
      json
      redis (>= 3.0.6)
      redis-namespace (>= 1.3.1)
    sidekiq-pro (1.9.2)
      sidekiq (>= 3.2.3)
```

**Comment by mperham** [comment]
show us your queue configuration.

**Comment by parhamr** [comment]
Configurations…

* `config/sidekiq.yml`

    ```
    background:
      :verbose: false
      :environment: background
      # NOTE: timeout deprecated as it has architectural flaws; kept here for process signaling integrations
      :concurrency: 8
      # timeout must be less than the time bluepill waits to kill (pause after USR1)
      :timeout: 85
      :queues:
        - [default, 3]
        - [mailer, 7]
        - [serializers, 1]
        - [admin, 5]
        - [reports, 1]
        - [messages, 3]
      :redis:
        url: redis://10.54.164.133:6379/2
    ```

* `config/initializers/sidekiq.rb`

    ```
    sidekiq_yml = 'config/sidekiq' + (ENV['SIDEKIQ_YAML'].blank? ? '' : ('_' + ENV['SIDEKIQ_YAML'])) + '.yml'
    
    # https://github.com/mperham/sidekiq/wiki/Reliability
    require 'sidekiq/pro/reliable_push'
    
    Sidekiq.configure_server do |config|
      require 'sidekiq/pro/reliable_fetch'
      sidekiq_config = AppName::Utils.load_environment_config_file(Rails.root.join(sidekiq_yml))
      config.options[:verbose]      = sidekiq_config[:verbose]
      puts sidekiq_config[:queues].inspect
      #config.options[:queues]       = sidekiq_config[:queues]
      config.options[:environment]  = sidekiq_config[:environment] || 'background'
      config.options[:concurrency]  = sidekiq_config[:concurrency] || 10
      config.options[:logfile]      = sidekiq_config[:logfile]
      config.options[:pidfile]      = sidekiq_config[:pidfile]
      # timeout must be less than the time bluepill waits to kill (pause after USR1)
      config.options[:timeout]      = sidekiq_config[:timeout] || 90
      config.redis                  = sidekiq_config[:redis].merge(namespace: 'sidekiq')
    
      config.server_middleware do |chain|
        chain.add Kiqstand::Middleware
        chain.add Sidekiq::Middleware::Server::Batch
      end
      config.client_middleware do |chain|
        chain.add Sidekiq::Middleware::Client::Batch
      end
    end
    Sidekiq.configure_client do |config|
      sidekiq_config = AppName::Utils.load_environment_config_file(Rails.root.join(sidekiq_yml))
      # namespace must be defined here, per @mperham
      config.redis   = sidekiq_config[:redis].merge(namespace: 'sidekiq')
      config.client_middleware do |chain|
        chain.add Sidekiq::Middleware::Client::Batch
      end
    end
    ```

**Comment by parhamr** [comment]
STDOUT content during startup:

```
2015-02-13T00:24:03.109Z 8429 TID-40ftc INFO: [Sidetiq] Sidetiq v0.6.3 - Copyright (c) 2012-2013, Tobias Svensson <tob@tobiassvensson.co.uk>
2015-02-13T00:24:03.109Z 8429 TID-40ftc INFO: [Sidetiq] Sidetiq is covered by the 3-clause BSD license.
2015-02-13T00:24:03.109Z 8429 TID-40ftc INFO: [Sidetiq] See LICENSE and http://opensource.org/licenses/BSD-3-Clause for licensing details.
2015-02-13T00:24:03.109Z 8429 TID-40ftc INFO: [Sidetiq] Sidetiq::Supervisor start
2015-02-13T00:24:03.122Z 8429 TID-nwmnk INFO: [Sidetiq] Sidetiq::Actor::Clock id: 40155160 initialize
2015-02-13T00:24:03.142Z 8429 TID-nvfd8 INFO: [Sidetiq] Sidetiq::Actor::Handler id: 40113740 initialize
2015-02-13T00:24:03.152Z 8429 TID-nurlo INFO: [Sidetiq] Sidetiq::Actor::Handler id: 40077280 initialize
[["default", 3], ["mailer", 7], ["serializers", 1], ["admin", 5], ["reports", 1], ["messages", 3]]
2015-02-13T00:24:05.185Z 8429 TID-40ftc INFO: Booting Sidekiq 3.2.6 with redis options {:url=>"redis://10.54.164.133:6379/2", :namespace=>"sidekiq"}
2015-02-13T00:24:08.664Z 8429 TID-40ftc INFO: Running in ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-linux]
2015-02-13T00:24:08.664Z 8429 TID-40ftc INFO: Sidekiq Pro 1.9.2, commercially licensed.  Thanks for your support!
2015-02-13T00:24:08.664Z 8429 TID-40ftc INFO: Starting processing, hit Ctrl-C to stop
2015-02-13T00:24:08.812Z 8429 TID-da6ro INFO: ReliableFetch activated
2015-02-13T00:24:08.825Z 8429 TID-da6ro WARN: ReliableFetch: recovering work on 7 messages
```

**Comment by parhamr** [comment]
If it further helps, here’s a New Relic comparison of the same, unchanged worker class behaviors…

Sidekiq Pro 1.9.1, captured on Jan 16, 2015:

![screen shot 2015-02-12 at 4 47 32 pm](https://cloud.githubusercontent.com/assets/177123/6180245/fcab5204-b2d6-11e4-99e4-f159fb3a27e6.png)

------

Sidekiq Pro 1.9.2, captured on Feb 12, 2015:

![screen shot 2015-02-12 at 4 46 26 pm](https://cloud.githubusercontent.com/assets/177123/6180251/0f150b6a-b2d7-11e4-88e9-9033a08c7da5.png)

FYI: this isn’t an urgent issue… enjoy the wonderful weather! :sunglasses:

**Comment by mperham** [comment]
There's no change that should increase worker runtime.  The only fix to 1.9.2 changes Sidekiq Pro's startup, not runtime.  I suspect you've got another change which caused the problem.  Did you upgrade other gems at the same time?

**Comment by parhamr** [comment]
Yes, I did upgrade other gems. I'll check them out for the regression source. Thanks for confirming!
